Add support for MTP class device

roundby opened the pull request:
**Describe the PR**
Add Media Transfer Protocol (MTP) device class support

**Additional context**
Feature request reference: https://github.com/hathach/tinyusb/issues/1079

The implementation has been tested and found functional at prototype level. We provide an example tested on STM32F3. It makes use of blulk and interrupt EPs only, so it should be supported by most platforms.

The class supports a single session only, but it allows multiple objects and nested associations (i.e. files and folders). We did not add support for object properties access. There are some other limitations but for the purpose of exposing file trees when mass storage class is not advised, the library is fully functional.

This implementation has been written from the ground up, with minimal requirements on buffers and keeping the code as simple as possible. Most of the buffer allocation for object I/O is demanded to the underlying implementation, as such, the RAM footprint of the library is limited.

The working mechanism of MTP (persistent 32 bit file handles) may not scale well on microcontrollers with constrained resources, depending on the application requirements. In the example, we just provide a fixed filesystem with files residing in RAM. It would be possible, depending on the application requirements, to build filesystem based I/O on top of this, by using the API provided in mtp_device.h.

The sources are available [here](https://github.com/ennebi/tinyusb/tree/mtp)

WerWolv commented:
Congrats! 
I started working on my own MTP driver about a week ago and just got the first few things working now before I found your PR. Nowhere near done yet but here's my branch in case it helps anybody: https://github.com/WerWolv/tinyusb/tree/sicd

roundby commented:
@WerWolv thank you for sharing!
I am giving some context here, as it might be beneficial for review. I had a short look at your code, the implementation is not that far from the one we're suggesting. Since you already have a hands-on experience on how the MTP class works, it would be great if you could take a look at our code too (feel free to add comments or propose corrections) and maybe perform a test on a physical device (since now, we tested it on STM32F3 and on a custom PSoC5LP port, on Windows hosts). Hopefully, the example should work out of the box on most of the already supported demo boards.

This code has been designed after the need to expose a set of files and folders stored in an SPI memory managed by littlefs, and we took the chance to do it within this library (we already used TinyUSB on some of our designs, and we hope this is a way to contribute back). While reading the code and the example, one should remind that the specification requires the device to provide a session unique 32 bit identifier for each object (e.g. file or directory), cached by the host when traversing storage content, able to be recalled any time and never re-used if an object is deleted. The implementation of this on the application side and in case of static files is easy, but it's not trivial on mutable file system where file I/O is performed by both host and (internally) by the device. Moreover, we found that, under Windows (appearing in contrast with MTP requirements) the object IDs are still cached if the session is closed and reopened, unless a bus reset or physical unplug is performed.

Comparing with your code:

- mtp.h contains some of the definitions you have in sicd.h. We added packed data structures for fixed data as we found this easier to read and similar to what is already done (see mass storage device class).
- We provided application callbacks in mtp_device_storage.h, in order for the application to provide the interface to enumerate, read and write objects. Comments should be self-explanatory, but feel free to ask in case of doubts.
- mtp_device.c contains the actual implementation (similar to your sicd_device.c). We did not require a minimum EP size, as the buffer I/O is managed by the application (usually the filesystem itself), but we statically allocated a single buffer with size MTP_MAX_PACKET_SIZE (512). With this single buffer, the code supports arbitrary object sizes. Each command is handled in a single function and information is taken from tud_mtp_* callbacks.
- The example code is under examples/device/mtp/src/mtp_fs_example.c. A few MTP class specific definitions, namely CFG_MTP_*, are added to tusb_config.h.
- We did not add support for properties (we don't needed them, and they would have increased the amount of code to review). 
 
The code is functional, and it is expected to be ready for exposing file and directory level I/O when a block device is not advisable:

- Storage is visible in Explorer
- Objects (files and directories) can be enumerated
- Object reading, creation and deletion works

I recognize that the introduction of a new class (and it's the first time we're contributing to TinyUSB) could require some time for a review and refinement, so, we'll wait for indications from the maintainers about the interest in merging this, asking for changes, moving to a different branch for testing/integration with additional features or any other suggestion. So far, we corrected all the notified build errors and most of the code scanning notices.

simon88 commented:
Hi @roundby 
Great work, do you try it with littlefs?

f-hoepfinger-hr-agrartechnik commented:
> Hi @roundby Great work, do you try it with littlefs?

littlefs must work for us, to be useful. 
Thanks for this great work!

f-hoepfinger-hr-agrartechnik commented:
> Hi @roundby Great work, do you try it with littlefs?

i guess it will be hard with littlefs. 
easier to make 1 FAT Partition, and copy the stuff over to littlefs later.

diancity025 commented:
Hi, I'm having some issues testing the features from https://github.com/ennebi/tinyusb/tree/mtp, copying or creating files in Windows 10 doesn't work properly.
I printed some logs and found that the mtpd_handle_cmd_send_object_info function was called when the file was created, and the mtpd_handle_dto_send_object_info function was not called to write after that.
How do I solve the problem? These are my test environments:
Hardware: RP2040 SDK2.1.1

roundby commented:
>I printed some logs and found that the mtpd_handle_cmd_send_object_info function was called when the file was created, and the mtpd_handle_dto_send_object_info function was not called to write after that.<

Hi @diancity025.
Is the example code (read only FS) working?


Can you check if the bulk endpoints are numbered and opened correctly in the initialization sequence? The command phase is supposed to be followed by a data phase on bulk EP, unless there is an issue with the sequence.
Would you have the chance to share your code? I could check it on a RBPI in a while.

diancity025 commented:
Hi @roundby Read-only works normally. The problem I encountered before could not create files and folders in WIN10. I added some changes through debugging and now it works fine. I'm not sure if this is a BUG.
[Uploading DIFF.PATCH…]()

roundby commented on examples/device/mtp/src/mtp_fs_example.c:
```diff
@@ -0,0 +1,315 @@
+    return MTP_RESC_OK;
+}
+
+mtp_response_t tud_mpt_storage_format(uint32_t storage_id)
```
corrected, thank you!

roundby commented:
Hi @f-hoepfinger-hr-agrartechnik 

> littlefs must work for us, to be useful.
> i guess it will be hard with littlefs. easier to make 1 FAT Partition, and copy the stuff over to littlefs later.

Lately, I am replying to this.

We used this change in an internal test project, specifically designed to expose files from an SPI memory managed by littlefs. I can confirm that this setup is perfectly feasible, though there are some limitations depending on platform constraints.

You can imagine that including a complete example with littlefs as a dependency would be too heavy for many microcontrollers, and likely wouldn’t be accepted. However, implementation is straightforward—as long as you account for the system's limitations.

One option is to implement a flat filesystem (i.e., avoid creating associations), storing the unique identifier in either the file name or file metadata. If the platform has sufficient RAM, you could instead track queried objects and construct a mapping between identifiers and file paths. Of course there might be other valid approaches.

This mapping strategy is likely the only major design decision you'll need to make. Everything else should be relatively straightforward.

diancity025 commented:
Hi @roundby, thanks very much for this!
    I recently had an issue where I couldn't store large files when I was using RP2040 to test mtp docking with littlefs, I saved the file handle via the LFS_TYPE_USERATTR attribute of the littlefs file. At the moment it is normal when reading and writing small files. But when I write to a relatively large file (368k), I have the problem that the transfer timeout cannot be completed. Can you help with that? Thank you.
Here are the logs I scraped.
```
Device  Phase  Data                      Description       Cmd.Phase.Ofs(rep)
------  -----  ------------------------  ----------------  ------------------
  41.1  OUT    14 00 00 00  01 00 0c 10  ........                 1.1.0        
               17 00 00 00  01 00 01 00  ........                 1.1.8        
               ff ff ff ff               ....                     1.1.16       
  41.1  OUT    a4 00 00 00  02 00 0c 10  ........                 2.1.0        
               17 00 00 00  00 00 00 00  ........                 2.1.8        
               00 30 00 00  29 05 06 00  .0..)...                 2.1.16       
               00 30 00 00  00 00 00 00  .0......                 2.1.24       
               00 00 00 00  00 00 00 00  ........                 2.1.32       
               00 00 00 00  00 00 00 00  ........                 2.1.40       
               00 00 00 00  00 00 00 00  ........                 2.1.48       
               00 00 00 00  00 00 00 00  ........                 2.1.56       
  41.1  OUT    0c 68 00 6f  00 6c 00 64  .h.o.l.d                 3.1.0        
               00 32 00 32  00 77 00 2e  .2.2.w..                 3.1.8        
               00 7a 00 69  00 70 00 00  .z.i.p..                 3.1.16       
               00 12 32 00  30 00 32 00  ..2.0.2.                 3.1.24       
               35 00 30 00  37 00 30 00  5.0.7.0.                 3.1.32       
               35 00 54 00  31 00 30 00  5.T.1.0.                 3.1.40       
               35 00 35 00  35 00 31 00  5.5.5.1.                 3.1.48       
               2e 00 30 00  00 00 12 32  ..0....2                 3.1.56       
  41.1  OUT    00 30 00 32  00 34 00 31  .0.2.4.1                 4.1.0        
               00 32 00 31  00 30 00 54  .2.1.0.T                 4.1.8        
               00 31 00 37  00 33 00 34  .1.7.3.4                 4.1.16       
               00 34 00 38  00 2e 00 30  .4.8...0                 4.1.24       
               00 00 00 00               ....                     4.1.32       
  41.2  IN     18 00 00 00  03 00 01 20  .......                  5.1.0        
               17 00 00 00  01 00 01 00  ........                 5.1.8        
               00 00 00 00  0a 00 00 00  ........                 5.1.16       
  41.1  OUT    0c 00 00 00  01 00 0d 10  ........                 6.1.0        
               18 00 00 00               ....                     6.1.8        
  41.1  OUT    35 05 06 00  02 00 0d 10  5.......                 7.1.0        
               18 00 00 00  50 4b 03 04  ....PK..                 7.1.8        
               0a 00 00 00  08 00 60 6f  ......`o                 7.1.16       
               73 59 e0 b0  c8 9b da 00  sY......                 7.1.24       
               00 00 53 01  00 00 16 00  ..S.....                 7.1.32       
               09 00 45 76  65 6e 74 52  ..EventR                 7.1.40       
               65 63 6f 72  64 65 72 53  ecorderS                 7.1.48       
               74 75 62 2e  73 63 76 64  tub.scvd                 7.1.56       
  41.0  CTL    a1 67 00 00  02 00 24 00  GET DEVICE STS           8.1.0(10)    
  41.0  IN     04 00 19 20               ...                      8.2.0        
  41.0  CTL    21 64 00 00  02 00 06 00  CANCEL                  18.1.0        
  41.0  USTS   c0010000                  canceled                18.2.0        
  41.0  CTL    a1 67 00 00  02 00 24 00  GET DEVICE STS          19.1.0        
  41.0  IN     04 00 19 20               ...                     19.2.0        
  41.1  OUT    10 00 00 00  01 00 05 10  ........                20.1.0        
               19 00 00 00  01 00 01 00  ........                20.1.8        
  41.2  IN     0c 00 00 00  03 00 01 20  .......                 21.1.0        
               18 00 00 00               ....                    21.1.8        
```

hhuynh222 commented:
> Hi @f-hoepfinger-hr-agrartechnik
> 
> > littlefs must work for us, to be useful.
> > i guess it will be hard with littlefs. easier to make 1 FAT Partition, and copy the stuff over to littlefs later.
> 
> Lately, I am replying to this.
> 
> We used this change in an internal test project, specifically designed to expose files from an SPI memory managed by littlefs. I can confirm that this setup is perfectly feasible, though there are some limitations depending on platform constraints.
> 
> You can imagine that including a complete example with littlefs as a dependency would be too heavy for many microcontrollers, and likely wouldn’t be accepted. However, implementation is straightforward—as long as you account for the system's limitations.
> 
> One option is to implement a flat filesystem (i.e., avoid creating associations), storing the unique identifier in either the file name or file metadata. If the platform has sufficient RAM, you could instead track queried objects and construct a mapping between identifiers and file paths. Of course there might be other valid approaches.
> 
> This mapping strategy is likely the only major design decision you'll need to make. Everything else should be relatively straightforward.

Can confirm this works with littlefs. Made some adjustments but can read all files from NOR flash

jtherin commented:
@hathach ? :)

roundby commented:
> I recently had an issue where I couldn't store large files when I was using RP2040 to test mtp docking with littlefs, I saved the file handle via the LFS_TYPE_USERATTR attribute of the littlefs file. At the moment it is normal when reading and writing small files. But when I write to a relatively large file (368k), I have the problem that the transfer timeout cannot be completed. Can you help with that? Thank you. Here are the logs I scraped.
> 

Hi @diancity025, thank you for reporting. I made a series of tests on RBPi Pico2 (RP2350) and I was able to replicate the issue only if verbose console output is enabled. From my understanding, Windows has a sort of fixed timeout over data chunks of 1MB, even if data transfer is in progress. By ensuring 1MB is transferred in less than 30 seconds I was able to transfer (both tx and rx) data sizes in the order of the hundreds of MBs without issues.

I also updated the code to clean the session once it is closed, and some other small fixes to operate under libmtp. Even with verbose debug output enabled, libmtp transfer the file correctly when Windows Explorer fails.

hathach commented:
mtp example enumerate, but I got permission issue when trying to open the readme.txt on debian 12, not sure what is wrong. Btw, I also try to make modification to this PR but it seems that I couldn't push to the fork. @roundby let me know if you could give me the write permission to your PR's branch, or we can merge this as it is since it look good enough and I could make an follow up PR to update later.  
![Uploading image.png…]()

hathach reviewed (commented):
I have added all the constant enum in the MTP specs, and do some rename as well as sync with master. Haven't made any behavior changes moification juts yet. I will try to troubleshoot the issue, please make sure you pull before doing more changes. Great PR so far, thank you.

hathach commented:
@roundby thank you, I am glad the refactor make sense to you. Indeed I think we should let application take full control of the data/response. Since it will be different from app to apps. Some device will has slow block storage such as SDcard and will only need to call the tud_mtp_data_send() when it got the data asynchronously. Actually the driver can handle part of the request e.g get_device_info() the data is pre-filled according to the CFG_TUD_MTP_DEVICE_INFO_* but still presented the whole buffer to application, in case it want to do the full changes. Like msc, I think the driver shouldd focus on making reading/writing large data (e.g few MBs) easier for application. Still working on more update, hopefully we can merge this soon enough.

hathach reviewed (approved):
I have completed the rework, including adding support control request. I found that most command is rather application specific with filesystem manipulation. Therefore I think it would be easier for both the stack (and application) to forward most of the command handler to user. Though we can pre-process some command such as get device info (still the whole container is given to user should they want to have their own from sratch). This can change/expand in the future though. @roundby let me know if the update version make sense to you and we can merge this.

BenBE commented on examples/device/mtp/src/usb_descriptors.c:
```diff
@@ -109,18 +109,79 @@ enum
+// Per USB specs: high speed capable device must report device_qualifier and other_speed_configuration
+
+// high speed configuration
+uint8_t const desc_hs_configuration[] = {
```
```suggestion
const uint8_t desc_hs_configuration[] = {
```

Why not consistent with the order used for the FS configuration above?

BenBE commented on examples/device/mtp/src/usb_descriptors.c:
```diff
@@ -109,18 +109,79 @@ enum
+  (void) index; // for multiple configurations
+
+  // if link speed is high return fullspeed config, and vice versa
+  // Note: the descriptor type is OHER_SPEED_CONFIG instead of CONFIG
```
```suggestion
  // Note: the descriptor type is OTHER_SPEED_CONFIG instead of CONFIG
```

hathach reviewed (approved):
I have added MTP to hil test, which test the device info: serial/manufacture/friendly name and file read write readme.txt logo.png and send file 1-3kb randomly. All works rather well. I think this PR is good to merge. @roundby should there is any issues, we can fix it in follows up
